### PR TITLE
Add network classes

### PR DESCRIPTION
**Pull request by maybegreat48**
Should I split the files further?

### CODE REVIEW

**Comment by Yimura** [comment]
Looks good, don't think it needs to be split up any more, though I'm open for comments from others.

The network class requires +0x70 on Steam builds of the game I believe

**Comment by SkiddyToast** [comment]
Can confirm it is +0x70 on steam, although that can be fixed after pattern scanning instead of changing the class.

**Comment by SkiddyToast** [comment]
AFAIK it starts from the beginning, although I could be wrong.

**Comment by SkiddyToast** [comment]
The Network constructor is quite large lmao
```c++
char *__fastcall Network::Network(char *a1)
{
  char *v2; // rbx
  int v3; // esi
  char *v4; // rax
  int i; // ecx
  char *v6; // rbx
  int j; // esi
  char *v8; // rax
  int k; // ecx
  char *v10; // rbx
  int m; // esi
  int v12; // esi
  char *v13; // rbx
  __int64 v14; // rbp
  int v15; // ecx
  char *v16; // rax
  char *v17; // rbx
  int n; // esi
  int v19; // r15d
  int v20; // esi
  char *v21; // rbx
  __int64 v22; // rcx
  __int64 v23; // rax
  int v24; // r9d
  char *v25; // rax
  _BYTE *v26; // rcx
  _QWORD *v27; // rdx
  __int64 v28; // r8
  int v29; // ecx
  char *v30; // rax
  int v31; // ecx
  char *v32; // rax
  char *v33; // rbx
  int ii; // esi
  int v35; // eax
  int v36; // eax
  int v37; // edx
  int v38; // esi
  char *v39; // rbx
  int jj; // r14d
  char *v41; // rbx
  char *v42; // rsi
  int v43; // ebx
  int v44; // r14d
  char *v45; // rsi
  char *v46; // rsi
  int kk; // r14d
  char *v48; // rsi
  int mm; // r14d
  char *v50; // rsi
  char *v51; // rax
  char *v52; // r8
  char *v53; // rdx

  rage::rlSessionInfo::rlSessionInfo(a1);
  v2 = a1 + 184;
  dword_141E43C30 = 214013 * dword_141E43C30 + 2531011;
  *((_DWORD *)a1 + 30) = dword_141E43C30;
  dword_141E43C30 = 214013 * dword_141E43C30 + 2531011;
  *((_DWORD *)a1 + 31) = dword_141E43C30;
  dword_141E43C30 = 214013 * dword_141E43C30 + 2531011;
  *((_DWORD *)a1 + 34) = dword_141E43C30;
  dword_141E43C30 = 214013 * dword_141E43C30 + 2531011;
  *((_DWORD *)a1 + 35) = dword_141E43C30;
  v3 = 1;
  *((_QWORD *)a1 + 20) = 0i64;
  do
  {
    rage::snSession::snSession(v2);
    v2 += 15984;
    --v3;
  }
  while ( v3 >= 0 );
  v4 = a1 + 32152;
  for ( i = 1; i >= 0; --i )
  {
    *(_QWORD *)v4 = 0i64;
    v4 += 8;
  }
  v6 = a1 + 32168;
  for ( j = 1; j >= 0; --j )
  {
    NetworkGameConfig::NetworkGameConfig((__int64)v6);
    v6 += 328;
  }
  v8 = a1 + 32828;
  for ( k = 1; k >= 0; --k )
  {
    *(_QWORD *)v8 = 0i64;
    v8 += 8;
  }
  MetricSessionMigrated::MetricSessionMigrated((__int64)(a1 + 32888));
  a1[33704] = 1;
  *((_DWORD *)a1 + 8427) = 0;
  NetworkGameFilter::NetworkGameFilter((__int64)(a1 + 33712));
  a1[34548] &= ~1u;
  *((_DWORD *)a1 + 8644) = -1;
  *((_DWORD *)a1 + 8636) = 0;
  *((_QWORD *)a1 + 4319) = 0i64;
  *((_QWORD *)a1 + 4320) = 0i64;
  *((_QWORD *)a1 + 4321) = 1067030938i64;
  *((_DWORD *)a1 + 8645) = 0;
  *((_WORD *)a1 + 17292) = 0;
  a1[34586] = 1;
  rage::rlSessionInfo::rlSessionInfo((_QWORD *)a1 + 4324);
  *((_DWORD *)a1 + 8682) = -1;
  a1[34704] = 0;
  *((_DWORD *)a1 + 8677) = 32;
  a1[34732] = 0;
  sub_1413C3C7C(a1 + 34736);
  *((_DWORD *)a1 + 8688) = 0;
  sub_1413C3C7C(a1 + 34760);
  sub_1411498CC(a1 + 34776);
  *((_DWORD *)a1 + 8816) = 0;
  *((_DWORD *)a1 + 8829) = 6;
  *((_WORD *)a1 + 17642) = 0;
  a1[35286] = 0;
  *((_QWORD *)a1 + 4411) = 6i64;
  *((_QWORD *)a1 + 4412) = 0i64;
  *((_DWORD *)a1 + 8826) = 0;
  *((_DWORD *)a1 + 8827) = 16777472;
  a1[35312] = 0;
  *((_DWORD *)a1 + 8830) = 4;
  a1[35324] = 0;
  *((_DWORD *)a1 + 8832) = 10;
  *((_DWORD *)a1 + 8833) = 4;
  *((_DWORD *)a1 + 8834) = 1;
  v10 = a1 + 35344;
  for ( m = 1; m >= 0; --m )
  {
    `vector constructor iterator'(v10, 0x18ui64, 32, sub_14114D5B4);
    *((_DWORD *)v10 + 192) = 0;
    v10 += 776;
  }
  *((_DWORD *)a1 + 9224) = 30000;
  a1[36900] = 1;
  v12 = 1;
  v13 = a1 + 37420;
  do
  {
    `vector constructor iterator'(v13 - 516, 0x10ui64, 32, sub_1413C3C7C);
    --v12;
    *(_QWORD *)(v13 - 4) = 0i64;
    v13 += 520;
  }
  while ( v12 >= 0 );
  *((_WORD *)a1 + 18972) = 0;
  *((_DWORD *)a1 + 9487) = 0;
  sub_1413C5150(a1 + 37960);
  rage::rlSessionInfo::rlSessionInfo((_QWORD *)a1 + 4864);
  v14 = 2i64;
  *(_QWORD *)(a1 + 39028) = 0i64;
  *(_QWORD *)(a1 + 39036) = 0i64;
  v15 = 2;
  v16 = a1 + 39052;
  *((_WORD *)a1 + 19522) = 0;
  do
  {
    --v15;
    *(_DWORD *)v16 = 0;
    v16[4] = 0;
    v16 += 8;
  }
  while ( v15 >= 0 );
  *((_DWORD *)a1 + 9769) = -1;
  *((_WORD *)a1 + 19556) = 255;
  *((_DWORD *)a1 + 9779) = 0;
  *((_WORD *)a1 + 19560) = 0;
  *((_DWORD *)a1 + 9781) = 0;
  a1[39128] = 0;
  *((_DWORD *)a1 + 9863) = 4;
  a1[39456] = 1;
  *(_QWORD *)(a1 + 39460) = 1065353216i64;
  *((_DWORD *)a1 + 9867) = 5;
  *((_DWORD *)a1 + 9868) = 1036831949;
  *((_DWORD *)a1 + 9869) = 1069547520;
  v17 = a1 + 39488;
  for ( n = 2; n >= 0; --n )
  {
    sub_14114D498(v17);
    v17 += 14448;
  }
  v19 = 9;
  *((_QWORD *)a1 + 10354) = 0i64;
  *((_DWORD *)a1 + 20710) = 0;
  *((_QWORD *)a1 + 10356) = 0i64;
  *((_QWORD *)a1 + 10357) = 0i64;
  a1[82864] = 0;
  v20 = 9;
  v21 = a1 + 82872;
  do
  {
    sub_1413C5150(v21);
    v21 += 976;
    --v20;
  }
  while ( v20 >= 0 );
  *((_DWORD *)a1 + 23158) = 0;
  a1[92636] = 0;
  *((_DWORD *)a1 + 23161) = 0;
  *((_DWORD *)a1 + 23168) = 15000;
  *((_WORD *)a1 + 46338) = 257;
  a1[92678] = 1;
  *((_QWORD *)a1 + 11585) = 0i64;
  *((_QWORD *)a1 + 11586) = 0i64;
  sub_1417D9560(a1 + 92704, 0i64, 80i64);
  *((_DWORD *)a1 + 23196) = 0;
  *((_QWORD *)a1 + 11599) = 0i64;
  *((_QWORD *)a1 + 11600) = 0i64;
  *((_DWORD *)a1 + 23202) = 0;
  a1[92812] = 0;
  *((_QWORD *)a1 + 11602) = 0i64;
  *((_QWORD *)a1 + 11603) = 0i64;
  a1[92832] = 0;
  *(_DWORD *)(a1 + 92834) = 0;
  *((_DWORD *)a1 + 23210) = 0;
  *((_WORD *)a1 + 46422) = 0;
  *((_QWORD *)a1 + 11606) = 0i64;
  a1[92856] = 0;
  *(_DWORD *)(a1 + 92858) = 0;
  *((_WORD *)a1 + 46431) = 0;
  *((_QWORD *)a1 + 11608) = 65280i64;
  *((_WORD *)a1 + 46436) = 0;
  *(_DWORD *)(a1 + 92878) = 0;
  *((_WORD *)a1 + 46438) = 0;
  *((_DWORD *)a1 + 23229) = 0;
  a1[92914] = 4;
  *(_QWORD *)(a1 + 92882) = -1i64;
  *(_QWORD *)(a1 + 92890) = -1i64;
  *(_QWORD *)(a1 + 92898) = -1i64;
  *(_QWORD *)(a1 + 92906) = -1i64;
  v22 = 4i64;
  *((_WORD *)a1 + 46460) = 0;
  v23 = 48i64;
  do
  {
    *(_DWORD *)&a1[v23 + 92876] = 0;
    v23 += 4i64;
    --v22;
  }
  while ( v22 );
  *((_WORD *)a1 + 46470) = 0;
  a1[92944] = 0;
  sub_1413C3C7C(a1 + 92952);
  sub_141155164(a1 + 92952);
  *((_WORD *)a1 + 46496) = 0;
  a1[92994] = 0;
  v24 = 1;
  v25 = a1 + 93002;
  do
  {
    *(_DWORD *)(v25 - 2) = 256;
    v26 = v25 + 38;
    v27 = v25 + 6;
    v28 = 4i64;
    do
    {
      *v27 = 0i64;
      *v26++ = 0;
      ++v27;
      --v28;
    }
    while ( v28 );
    v25 += 48;
    --v24;
  }
  while ( v24 >= 0 );
  v29 = 1;
  v30 = a1 + 93136;
  do
  {
    --v29;
    *((_QWORD *)v30 - 5) = 0i64;
    *((_QWORD *)v30 - 4) = 0i64;
    *((_QWORD *)v30 - 3) = 0i64;
    *((_QWORD *)v30 - 2) = 0i64;
    *((_QWORD *)v30 - 1) = 0i64;
    *(_QWORD *)v30 = 0i64;
    *((_QWORD *)v30 + 1) = 0i64;
    *((_QWORD *)v30 + 2) = 0i64;
    v30 += 64;
  }
  while ( v29 >= 0 );
  v31 = 1;
  v32 = a1 + 93232;
  do
  {
    --v31;
    *((_QWORD *)v32 - 1) = 0i64;
    *(_QWORD *)v32 = 0i64;
    *((_QWORD *)v32 + 1) = 0i64;
    *((_QWORD *)v32 + 2) = 0i64;
    *((_QWORD *)v32 + 3) = 0i64;
    v32 += 40;
  }
  while ( v31 >= 0 );
  sub_14114CA0C(a1 + 93304);
  sub_14114CA0C(a1 + 93360);
  sub_14114CA0C(a1 + 93416);
  sub_1413C3C7C(a1 + 93488);
  v33 = a1 + 93504;
  for ( ii = 1; ii >= 0; --ii )
  {
    CNetComplaintMgr::CNetComplaintMgr(v33);
    v33 += 3224;
  }
  v35 = dword_142AFC60C;
  *((_QWORD *)a1 + 12494) = 0i64;
  *((_QWORD *)a1 + 12495) = 0i64;
  *((_DWORD *)a1 + 24992) = v35;
  v36 = dword_141F044C8;
  v37 = dword_141F044C8 >> 31;
  *((_WORD *)a1 + 49988) = 257;
  a1[99978] = 1;
  *((_DWORD *)a1 + 24993) = ((v36 - v37) >> 1) + 5000;
  sub_14114B3C4(a1 + 99984);
  sub_14114A8CC(a1 + 106432);
  *((_DWORD *)a1 + 26782) = 0;
  sub_14114985C(a1 + 107136);
  *((_WORD *)a1 + 55172) = 257;
  *((_QWORD *)a1 + 13794) = 0i64;
  *((_DWORD *)a1 + 27590) = 0;
  *((_QWORD *)a1 + 13796) = 0i64;
  *((_DWORD *)a1 + 27594) = 0;
  *((_DWORD *)a1 + 27596) = 0;
  *((_WORD *)a1 + 55194) = 0;
  *((_QWORD *)a1 + 13799) = 0i64;
  rage::rlSessionInfo::rlSessionInfo((_QWORD *)a1 + 13800);
  *((_DWORD *)a1 + 27629) = 0;
  *((_WORD *)a1 + 55260) = 0;
  *(_QWORD *)(a1 + 110524) = 0i64;
  *((_DWORD *)a1 + 27633) = 0;
  a1[110536] = 0;
  *((_WORD *)a1 + 55269) = 0;
  *((_DWORD *)a1 + 27635) = 0;
  *((_QWORD *)a1 + 13821) = 65792i64;
  *((_DWORD *)a1 + 27644) = 2;
  a1[110580] = 0;
  *((_QWORD *)a1 + 13823) = 0i64;
  *((_DWORD *)a1 + 27648) = 0;
  *((_WORD *)a1 + 55298) = 257;
  sub_1411499C0(a1 + 110624);
  sub_1413C3C7C(a1 + 112176);
  a1[112192] = 0;
  *((_QWORD *)a1 + 14025) = 0i64;
  *((_DWORD *)a1 + 28052) = 0;
  a1[112212] = 0;
  sub_1413CCD48(a1 + 112176);
  *((_DWORD *)a1 + 28054) = 0;
  *((_WORD *)a1 + 56110) = 0;
  *(_DWORD *)(a1 + 112785) = 256;
  a1[112789] = 1;
  *((_DWORD *)a1 + 28198) = 0;
  sub_1413C3C7C(a1 + 112800);
  a1[112816] = 0;
  v38 = 3;
  *((_DWORD *)a1 + 28205) = 0;
  v39 = a1 + 112824;
  for ( jj = 3; jj >= 0; --jj )
  {
    sub_14114D498(v39);
    v39 += 14448;
  }
  v41 = a1 + 170616;
  do
  {
    NetworkGameFilter::NetworkGameFilter((__int64)v41);
    v41 += 832;
    --v38;
  }
  while ( v38 >= 0 );
  v42 = a1 + 173984;
  *((_QWORD *)a1 + 21743) = 0i64;
  *((_DWORD *)a1 + 43488) = 0;
  *((_DWORD *)a1 + 43490) = 0;
  a1[173964] = 0;
  v43 = 31;
  a1[173976] = 0;
  v44 = 31;
  *((_DWORD *)a1 + 43495) = 0;
  do
  {
    sub_1413C3C7C(v42);
    v42 += 16;
    --v44;
  }
  while ( v44 >= 0 );
  a1[174497] = 0;
  rage::rlSessionInfo::rlSessionInfo((_QWORD *)a1 + 21813);
  a1[174616] = 0;
  *((_QWORD *)a1 + 21828) = 0i64;
  *((_DWORD *)a1 + 43658) = 0;
  *((_QWORD *)a1 + 21830) = 0i64;
  *((_QWORD *)a1 + 21831) = 0i64;
  a1[174656] = 0;
  v45 = a1 + 174664;
  do
  {
    sub_1413C5150(v45);
    v45 += 976;
    --v19;
  }
  while ( v19 >= 0 );
  *((_DWORD *)a1 + 46106) = 1;
  sub_14114985C(a1 + 184432);
  *((_DWORD *)a1 + 46910) = 0;
  a1[187644] = 0;
  sub_1413C3C7C(a1 + 187656);
  rage::rlSessionInfo::rlSessionInfo((_QWORD *)a1 + 23459);
  *((_DWORD *)a1 + 46947) = 0;
  v46 = a1 + 187792;
  for ( kk = 31; kk >= 0; --kk )
  {
    sub_1413C3C7C(v46);
    v46 += 16;
  }
  *((_DWORD *)a1 + 47080) = 0;
  a1[188328] = 1;
  sub_1413C3C7C(a1 + 188336);
  sub_1413C3C7C(a1 + 188352);
  *((_DWORD *)a1 + 47092) = -1;
  *((_WORD *)a1 + 94186) = 256;
  a1[188374] = 0;
  v48 = a1 + 188376;
  for ( mm = 31; mm >= 0; --mm )
  {
    sub_1413C3C7C(v48);
    v48 += 16;
  }
  *((_DWORD *)a1 + 47222) = 0;
  v50 = a1 + 189008;
  a1[188892] = 0;
  do
  {
    sub_1413C3C7C(v50);
    v50 += 16;
    --v43;
  }
  while ( v43 >= 0 );
  *((_QWORD *)a1 + 18) = a1 + 184;
  *((_QWORD *)a1 + 19) = a1 + 16168;
  *((_QWORD *)a1 + 21) = a1 + 32152;
  *((_QWORD *)a1 + 22) = a1 + 32160;
  *((_QWORD *)a1 + 11653) = 0i64;
  *((_QWORD *)a1 + 11654) = sub_141174BA0;
  *((_QWORD *)a1 + 11653) = sub_140004354(a1);
  *((_QWORD *)a1 + 11658) = 0i64;
  *((_QWORD *)a1 + 11659) = sub_1411751C4;
  *((_QWORD *)a1 + 11658) = sub_140004354(a1);
  qword_142DB7F20 = 0i64;
  qword_142DB7F28 = (__int64)sub_141176904;
  qword_142DB7F20 = (__int64)sub_140004354(a1);
  qword_142DAD9F8 = 0i64;
  qword_142DADA00 = (__int64)sub_141172A60;
  qword_142DAD9F8 = (__int64)sub_140004354(a1);
  *((_QWORD *)a1 + 11639) = 0i64;
  *((_QWORD *)a1 + 11640) = sub_141172B08;
  *((_QWORD *)a1 + 11639) = sub_140004354(a1);
  *((_QWORD *)a1 + 11637) = 0i64;
  *((_QWORD *)a1 + 11638) = sub_141172938;
  *((_QWORD *)a1 + 11637) = sub_140004354(a1);
  *((_QWORD *)a1 + 11643) = 0i64;
  *((_QWORD *)a1 + 11644) = sub_141162708;
  *((_QWORD *)a1 + 11643) = sub_140004354(a1);
  *((_QWORD *)a1 + 11641) = 0i64;
  *((_QWORD *)a1 + 11642) = (char *)&loc_14118555B + 1;
  *((_QWORD *)a1 + 11641) = sub_140004354(a1);
  *((_QWORD *)a1 + 11647) = 0i64;
  *((_QWORD *)a1 + 11648) = sub_141173880;
  *((_QWORD *)a1 + 11647) = sub_140004354(a1);
  *((_QWORD *)a1 + 11645) = 0i64;
  *((_QWORD *)a1 + 11646) = sub_141172988;
  *((_QWORD *)a1 + 11645) = sub_140004354(a1);
  *((_QWORD *)a1 + 11651) = 0i64;
  *((_QWORD *)a1 + 11652) = sub_141162BC8;
  *((_QWORD *)a1 + 11651) = sub_140004354(a1);
  *((_QWORD *)a1 + 11649) = 0i64;
  *((_QWORD *)a1 + 11650) = sub_1411855D4;
  *((_QWORD *)a1 + 11649) = sub_140004354(a1);
  *((_QWORD *)a1 + 13826) = 0i64;
  *((_QWORD *)a1 + 13827) = 0i64;
  *((_QWORD *)a1 + 12495) = sub_141173898;
  *((_QWORD *)a1 + 12494) = 0i64;
  *((_QWORD *)a1 + 12494) = sub_140004354(a1);
  *((_QWORD *)a1 + 4885) = 0i64;
  *((_QWORD *)a1 + 4886) = 0i64;
  *((_DWORD *)a1 + 9774) = 0;
  v51 = a1 + 39482;
  v52 = a1 + 32872;
  v53 = a1 + 93476;
  do
  {
    *((_DWORD *)v53 - 15158) = 0;
    v51[53990] = 0;
    *(_DWORD *)v53 = 0;
    *v51 = 1;
    *(v51 - 2) = 0;
    *(v51 - 6658) = 0;
    ++v51;
    *(_QWORD *)v52 = 0i64;
    v53 += 4;
    v52 += 8;
    --v14;
  }
  while ( v14 );
  *((_WORD *)a1 + 19523) = 257;
  a1[39048] = 1;
  sub_141184818(a1, 0i64, v52);
  sub_1413CD2EC(a1 + 34592);
  _InterlockedExchange((volatile __int32 *)a1 + 9758, 0);
  *((_DWORD *)a1 + 9759) = 0;
  if ( (unsigned __int8)sub_1413E7DFC(a1) )
  {
    sub_141400048(0i64, 0i64);
    sub_1413CD2EC(a1);
  }
  return a1;
}
```

File: network/snSession.hpp [hunk 1]
```diff
@@ -0,0 +1,168 @@
+#include "../rage/rlGamerInfo.hpp"
```
Comment (Yimura) [inline network/snSession.hpp:1]: `#pragma once` was missed here.

**Review by Yimura — CHANGES_REQUESTED** [review]

**Review by Yimura — APPROVED** [review]

**Pull request merged**